### PR TITLE
feat(eslint-plugin): add minimal built-in flat presets

### PR DESCRIPTION
**Pull request by SukkaW**
Add minimal flat config presets for those who are building their own eslint flat presets and want to use their own parsers and settings.

The usage would be like:

```js
// eslint.config.js
import eslintPluginNext from 'eslint-plugin-next';

export default [
  {
    files: [],
    plugins: {},
    settings: {},
    languageOptions: {},
    rules: {}
  },

  eslintPluginNext.flatConfigs.recommended,
  // or
  eslintPluginNext.flatConfigs.coreWebVitals,
]
```

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 5fa58fd5c77cbbe8dba9e3d271f7e6b28a2f9b23

`pnpm test test/unit/eslint-plugin-next/google-font-display.test.ts`

<details>
<summary>Expand output</summary>

  ● Test suite failed to run

    TypeError: Cannot read properties of undefined (reading 'google-font-display')

      3 | import { rules } from '@next/eslint-plugin-next'
      4 |
    > 5 | const NextESLintRule = rules['google-font-display']
        |                             ^
      6 |
      7 | const tests = {
      8 |   valid: [

      at Object.<anonymous> (unit/eslint-plugin-next/google-font-display.test.ts:5:29)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| buildDuration | 18.4s | 15.7s | N/A |
| buildDurationCached | 14.8s | 12.6s | N/A |
| nodeModulesSize | 410 MB | 410 MB | N/A |
| nextStartRea..uration (ms) | 472ms | 476ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| 1187-HASH.js gzip | 50.8 kB | 50.8 kB | N/A |
| 8276.HASH.js gzip | 169 B | 168 B | N/A |
| 8377-HASH.js gzip | 5.36 kB | 5.36 kB | N/A |
| bccd1874-HASH.js gzip | 53 kB | 53 kB | N/A |
| framework-HASH.js gzip | 57.5 kB | 57.5 kB | N/A |
| main-app-HASH.js gzip | 232 B | 235 B | N/A |
| main-HASH.js gzip | 34 kB | 34 kB | N/A |
| webpack-HASH.js gzip | 1.71 kB | 1.71 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 39.4 kB | 39.4 kB | ✓ |
| Overall change | 39.4 kB | 39.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 193 B | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 512 B | 510 B | N/A |
| css-HASH.js gzip | 343 B | 342 B | N/A |
| dynamic-HASH.js gzip | 1.84 kB | 1.84 kB | ✓ |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 363 B | 362 B | N/A |
| hooks-HASH.js gzip | 393 B | 392 B | N/A |
| image-HASH.js gzip | 4.49 kB | 4.49 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.35 kB | 2.34 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 328 B | ✓ |
| script-HASH.js gzip | 397 B | 397 B | ✓ |
| withRouter-HASH.js gzip | 323 B | 326 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.59 kB | 3.59 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| _buildManifest.js gzip | 749 B | 746 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| index.html gzip | 525 B | 523 B | N/A |
| link.html gzip | 539 B | 537 B | N/A |
| withRouter.html gzip | 520 B | 521 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| edge-ssr.js gzip | 128 kB | 128 kB | N/A |
| page.js gzip | 203 kB | 203 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 671 B | 667 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 31.2 kB | 31.2 kB | N/A |
| edge-runtime..pack.js gzip | 844 B | 844 B | ✓ |
| Overall change | 844 B | 844 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| 523-experime...dev.js gzip | 322 B | 322 B | ✓ |
| 523.runtime.dev.js gzip | 314 B | 314 B | ✓ |
| app-page-exp...dev.js gzip | 323 kB | 323 kB | ✓ |
| app-page-exp..prod.js gzip | 127 kB | 127 kB | ✓ |
| app-page-tur..prod.js gzip | 140 kB | 140 kB | ✓ |
| app-page-tur..prod.js gzip | 135 kB | 135 kB | ✓ |
| app-page.run...dev.js gzip | 313 kB | 313 kB | ✓ |
| app-page.run..prod.js gzip | 123 kB | 123 kB | ✓ |
| app-route-ex...dev.js gzip | 37.3 kB | 37.3 kB | ✓ |
| app-route-ex..prod.js gzip | 25.3 kB | 25.3 kB | ✓ |
| app-route-tu..prod.js gzip | 25.4 kB | 25.4 kB | ✓ |
| app-route-tu..prod.js gzip | 25.2 kB | 25.2 kB | ✓ |
| app-route.ru...dev.js gzip | 38.9 kB | 38.9 kB | ✓ |
| app-route.ru..prod.js gzip | 25.2 kB | 25.2 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.67 kB | 9.67 kB | ✓ |
| pages-api.ru...dev.js gzip | 11.6 kB | 11.6 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.67 kB | 9.67 kB | ✓ |
| pages-turbo...prod.js gzip | 21.7 kB | 21.7 kB | ✓ |
| pages.runtim...dev.js gzip | 27.4 kB | 27.4 kB | ✓ |
| pages.runtim..prod.js gzip | 21.7 kB | 21.7 kB | ✓ |
| server.runti..prod.js gzip | 916 kB | 916 kB | ✓ |
| Overall change | 2.36 MB | 2.36 MB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | SukkaW/next.js eslint-plugin-presets | Change |
| - | - | - | - |
| 0.pack gzip | 2.05 MB | 2.05 MB | ⚠️ +332 B |
| index.pack gzip | 71.6 kB | 73.4 kB | ⚠️ +1.77 kB |
| Overall change | 2.12 MB | 2.13 MB | ⚠️ +2.1 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 5fa58fd5c77cbbe8dba9e3d271f7e6b28a2f9b23

**Comment by SukkaW** [comment]
> TLDR: ESLint prefers not to use a `flatConfigs` export, but instead export all configs from `.configs`, and suffix old eslintrc-style configs with `-legacy`.

Adding a `-legacy` suffix means there will be a breaking change, which means we need to wait for another breaking change. Are you willing to wait for another year?

In the meantime, attaching an extra type-safe flat preset under the `configs` is impossible without triggering a TypeScript error. Self-reference also doesn't work, you will get a "used-before-declare" error.

**Comment by darthmaim** [comment]
Yeah I agree, just wanted to make sure that this is considered and that there is already a plan in place to match the recommendation in the next major next.js release.

Ideally the flat config would have been part of 15.0 along with eslint 9 support, sadly eslint 9 support landed with such short notice before 15.0 that no one catched this in canary :(

File: packages/eslint-plugin-next/src/index.ts [hunk 1]
```diff
@@ -25,39 +56,33 @@ module.exports = {
...
-      },
+      rules: coreWebVitalsRules,
+    },
+  },
+}
+
+module.exports = plugin
+module.exports.flatConfig = {
+  recommended: {
+    name: 'eslint-plugin-next recommended',
```
Comment (rakleed) [inline packages/eslint-plugin-next/src/index.ts:72]: It's probably better to call it `next/recommended`, like other plugins do?

- https://github.com/un-ts/eslint-plugin-import-x/blob/ac6d2e1b39e2a0d01b981b98b58aad0eb7939c70/src/index.ts#L161
- https://github.com/Rel1cx/eslint-react/blob/c7c18c6af30d6a26108a7d47aaa0da6bd56bd404/packages/plugins/eslint-plugin/src/configs/recommended.ts#L10
- https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/7566e13531f09a040daee4c16d0cba0c28d321c4/src/index.js#L303
- https://github.com/sindresorhus/eslint-plugin-unicorn/blob/1774135a5ddbded2c89f82952e37a3e3bb01cdfa/index.js#L73

Same for `core-web-vitals` config.

Comment (SukkaW) [reply packages/eslint-plugin-next/src/index.ts]: Done in [`8a7d85c` (#73873)](https://github.com/vercel/next.js/pull/73873/commits/8a7d85c1e791e6a0396d4592889c41c2e281ea59).

Comment (CHC383) [reply packages/eslint-plugin-next/src/index.ts]: @rakleed kindly ask to review and merge the changes, thanks

**Comment by theoludwig** [comment]
Related issue: https://github.com/vercel/next.js/issues/73655

**Review by rakleed — APPROVED** [review]

File: packages/eslint-plugin-next/src/index.ts [hunk 2]
```diff
@@ -25,39 +56,33 @@ module.exports = {
...
+    },
+    rules: recommendedRules,
+  },
+  coreWebVitals: {
+    name: 'next/core-web-vitals',
+    plugins: {
+      '@next/next': plugin,
+    },
+    rules: {
+      ...recommendedRules,
```
Comment (CHC383) [inline packages/eslint-plugin-next/src/index.ts:84]: Personally I think `coreWebVitals` should just include the `coreWebVitalsRules` instead of including the recommededRules, which is the same as the legacy configs, and users who want both presets could add them separately. However, I don't against the new rules set either, maybe it is up to the Next js members to decide.

**Review by CHC383 — APPROVED** [review]

**Comment by CHC383** [comment]
Hi @devjiwonchoi, I saw you were involved in #73655, could you review this PR? This doesn't fix #73655 but it is at least one step towards flat configs

**Review by juliolmuller — APPROVED** [review]

**Comment by juliolmuller** [comment]
I think this resolves #73655, #74133 and #75298.

**Review by devjiwonchoi — APPROVED** [review]

**Comment by devjiwonchoi** [comment]
Sorry for the delay, and thank you for the PR!

**Comment by Austin1serb** [comment]
> I think this resolves #73655, #74133 and #75298.

👋 Just confirming that #75298 is still reproducible with next@15.3.0-canary.18 and flat config using eslint@8.57.0. Here’s the output from running pnpm next lint:

```sql
Invalid Options:
- Unknown options: useEslintrc, extensions, resolvePluginsRelativeTo, rulePaths, ignorePath, reportUnusedDisableDirectives
- 'extensions' has been removed.
- 'resolvePluginsRelativeTo' has been removed.
- 'ignorePath' has been removed.
- 'rulePaths' has been removed. Please define your rules using plugins.
- 'reportUnusedDisableDirectives' has been removed. Please use the 'overrideConfig.linterOptions.reportUnusedDisableDirectives' option instead.
```
```vbnet
dependencies:
next 15.3.0-canary.18
eslint 8.57.0
```
